feat(i18n, draft): add new uiTranslator prop, add new UiTranslatorContext Provider

rexebin opened the pull request:
## Proposed changes

As per discussion on issue #865
1. add new `uiTranslator` prop, 
2. add new `UiTranslatorContext` Provider
3. apply translation as a quick trial to the new prop.

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

rexebin commented:
Not sure why the CI is failing. I have applied the `t` function to all labels I could find. Once settled, maybe we need to remove the existing `translations` prop in some components.

macrozone commented:
@rexebin thanks a lot, will review it today or tomorrow.

macrozone commented on docs/editor.md:
```diff
@@ -103,6 +103,31 @@ cellSpacing = {
+  'Move things': '移动',
+  'Resize things': '调整大小',
+  'Preview result': '预览模式',
+};
```
it would be good to have a type with all the translation keys.

rexebin replied:
The current structure with English words as the key is the laziest and simplest implement. Simplicity is also the main benefit of such design. It hands over 100% of responsibility to lib users and depending on requirements. The dictionary can be as partial or as complete as they want. 

Providing a complete dictionary for the lib users to translate is better, but it requires deep integrations. I have seen other libs do exactly this, most of them not bothering to provide any context descriptions, maybe because interface dictionaries are usually very small. Maybe later when there are contributions of dictionaries, react-page can switch to this solution.

As for using English words as key, I think it is only beneficial for dedicated keys when the same English words require different translations, if not happening, then separate keys posts no benefit because the English word and its translation pairs are unique. Of cause, unless I missed something.  I leave this to your good judgement.

I agree that add an extra description for describing the context could be beneficial. But with the current structure, many of them will be identical. For example, the labels for the buttons of Slate plugins, all labels go through the same render component `PluginButton`, context description needs to be added at a lower level in the plugin code. It means much deeper integrations for complete coverage.

How about changing the signature of `uiTranslator` to `(key: string, description?: string) => string`. The description will be given in English as the context in the source code as needed so that developers can catch it at their end. However, I think most of the labels are very simple and require no description.

Your existing translations object would require developers to provide it at every component as a prop, very performant and easy to use, but require a lot of work both for you and developers. If you agree with the current solution, I think removing `defaultTranslations` is the next step.

rexebin replied:
Maybe in the label, default to English words instead of solely rely on the `t` function. So that you have the flexibility to change labels and not worry about keys being changed.

macrozone replied:
fully agree! let's start simple. If users need more context, let's add something like the proposed `description` prop.

macrozone commented on examples/pages/empty.tsx:
```diff
@@ -15,6 +23,54 @@ const LANGUAGES: Options['languages'] = [
+    </div>
+  );
+}
+
```
please remove this from the "empty" example

rexebin replied:
Will do.

macrozone commented on packages/editor/src/core/Provider/index.tsx:
```diff
@@ -34,6 +34,7 @@ type ProviderProps = {
    * pass custom redux middleware:. Might get deprecated in the future
    */
   middleware?: Middleware[];
+  uiTranslator: (label?: string) => string | undefined;
```
can you add a jsdoc?

rexebin replied:
will do.

macrozone commented on packages/editor/src/core/components/hooks/__tests__/useNodeDropLevels.test.tsx:
```diff
@@ -50,7 +50,12 @@ const SAMPLE_STATE: Value = {
+        cellPlugins={[]}
+        lang="en"
+        value={SAMPLE_STATE}
+        uiTranslator={(label) => label}
```
i don't think that this test needs to be altered

rexebin replied:
The test is failing because `uiTranslator` is actually a required one for the `Provider`, I will change the Provider prop signature instead then.

macrozone replied:
i think it can be optional

macrozone commented on packages/editor/src/ui/BottomToolbar/NodeTools.tsx:
```diff
@@ -10,22 +10,22 @@ export type BottomToolbarMainBarProps = {
             <Grid item={true}>
               <Avatar
-                children={icon || (title ? title[0] : '')}
+                children={icon || t(title ? title[0] : '')}
```
the point here is that the avatar shows the first letter of the title if no icon is defined

so with translation it would be

`icon || t(title)[0]`

rexebin replied:
I think to remove `t` wrapper here may be better?

macrozone commented on packages/editor/src/ui/DisplayModeToggle/index.tsx:
```diff
@@ -42,14 +43,15 @@ export const DisplayModeToggle: React.SFC<{
+    { action: <ToggleInsert label={t(translations.insert)} /> },
+    { action: <ToggleLayout label={t(translations.layout)} /> },
+    { action: <ToggleResize label={t(translations.resize)} /> },
+    { action: <TogglePreview label={t(translations.preview)} /> },
```
this is one confusion that we have to discuss. Some stuff already has these translations. Should we keep them or should we migrate away?

rexebin replied:
The existing translations object would require developers to provide it at every component as a prop, very performant and easy to use, but require a lot of work both for you and developers. If you agree with the current solution, I think removing `translations` prop is the next step.

macrozone replied:
agreed!

macrozone reviewed (changes requested):
thank you very much for the work.

There are two things i am unsure about:

- what do we do with the existing translation mechanism that exists in some places?
- using the default english words as translation key is probably not a good idea. It needs some additional namespace, so that you at least know the context.

rexebin commented:
Done most. Now just need to rebase. also separate entry for `children` PR

rexebin commented:
merged, rebased, checked with dev server, everything is working. 
Struggled with rebase, guess I need to use rebase more, if not proper, please help and fix it. Usually, I only merge stuff one pushed to Github.

macrozone commented on packages/editor/src/core/types/node.ts:
```diff
@@ -98,3 +98,10 @@ export type CellDrag = {
+ */
+export type UiTranslator = {
+  t: (label?: string) => string | undefined;
+};
```
i would probably move it to the options.ts file instead of node.ts

rexebin replied:
Will do

macrozone commented on packages/editor/src/core/components/Cell/ErrorCell/index.tsx:
```diff
@@ -1,24 +1,27 @@
+          <dt>{t('Cause:') || 'Cause:'}</dt>
           <dd>{error.message}</dd>
-          <dt>Cell:</dt>
+          <dt>{t('Cell:') || 'Cell:'}</dt>
```
"don't repeat yourself"

we should remove the `|| 'text'` here. instead the t-function handles the cases where the translator function does return null (or an empty string).

Remember that we control the t function that is returned from `useUiTranslator`.

rexebin replied:
As we discussed before, the reason I added these `|| 'text'` is to show how we can separate `key` with labels. For example, if in the future, you need to change `Cell`, keeping the key untouched but add this instead. 

However, the default `(label) => label` defeat the purpose, we actually want to change the default `t` function to return `undefined` and hard code default labels in the code, just to separate keys with labels. Keys will never be changed, but labels could be changed. 

If you want to keep it simple, then default return keys as is would be a lot simpler, with fewer flexibilities.

We don't control the `t` function, we call it and use whatever been returned. Developers control this function.

Please let me know what you think of above.

macrozone replied:
> We don't control the t function, we call it and use whatever been returned. Developers control this function.

no, developers control the UiTranslator function.

I just really don't like to have now every string twice now in the code. It just adds noise to the code.

The t function should return a default translation if the developers UiTranslator returns null.

What about this proposel:

the `t`-functon (that geturns by the hook) has this implementation:

```
const t = (key: string) => {
 const translated = userDefinedUiTranslatorFunction(key)
 if(translated === null){
   return defaultTranslatorFunction(key) // that can be the identity to begin with and later can use some default translation object
 }
}
```

so it would always work, even if the developer forgets to translate something

macrozone commented on packages/editor/src/core/Provider/index.tsx:
```diff
@@ -137,23 +144,32 @@ const Provider: React.FC<ProviderProps> = ({
 
+  const uiTranslatorMemorized = useMemo(
+    () => ({
+      t: uiTranslator,
```
this function here should handle the case, where `uiTranslator` returns nothing. It should return a falllback value.

rexebin replied:
As I mentioned in the previews comment just above, depends on what you want, `uiTranslator` could just return key or `undefined`.

macrozone reviewed (changes requested):
see comments about duplicated strings

rexebin commented:
I am totally with you. Adding default labels repeatedly is a lot is not
good. If you don’t have issue with English words as key, then my initial
implementation defaulting to return key if not found is the simplest.

On Sun, 4 Apr 2021 at 12:17, Marco Wettstein ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In packages/editor/src/core/components/Cell/ErrorCell/index.tsx
> <https://github.com/react-page/react-page/pull/935#discussion_r606786813>:
>
> >            <dd>{error.message}</dd>
> -          <dt>Cell:</dt>
> +          <dt>{t('Cell:') || 'Cell:'}</dt>
>
> We don't control the t function, we call it and use whatever been
> returned. Developers control this function.
>
> no, developers control the UiTranslator function.
>
> I just really don't like to have now every string twice now in the code.
> It just adds noise to the code.
>
> The t function should return a default translation if the developers
> UiTranslator returns null.
>
> What about this proposel:
>
> the t-functon (that geturns by the hook) has this implementation:
>
> const t = (key: string) => {
>  const translated = userDefinedUiTranslatorFunction(key)
>  if(translated === null){
>    return defaultTranslatorFunction(key) // that can be the identity to begin with and later can use some default translation object
>  }
> }
>
> so it would always work, even if the developer forgets to translate
> something
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/react-page/react-page/pull/935#discussion_r606786813>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACTDVBYMWAYT5FG253MGL7TTHBDD3ANCNFSM4Z37S6BA>
> .
>

rexebin commented:
@macrozone Please see my latest commit. 
1. I have removed all default labels `|| 'text'`, they really weren't pretty. 
2. `UiTranslator` type is moved to `options.ts` as requested.
3.  changed `uiTranslator` prop from `(labels?: string) => string | undefined` to `(label:string) => string`. So that developers have to return a string if they are using TypeScript.

To recap: 
1. `uiTranslator` prop is optional, if not provided, `Editor` component default it to `label=>label` to the label as-is. It has very little impact on the rest of the code.
2.  English labels are used as keys. This assumes English labels are unique. If they are used in multiple places in `react-page`, their translations will be the same.
3. The signature `uiTranslator` prop will force developers to always return a string hopefully. If not, I guess they will find come blank labels quickly and fix the issue themselves.

We do have full control of the default `uiTranslator`. Later, if we have a complete set of translation for a language, it is very easy to introduce a `uiLanguage` for the supported language. All we have to do is inserting dictionaries into the default `uiTranslator` function to return the correct translated labels instead of returning labels as-is. We could even accept a partial dictionary from the developer and hook it into the default `uiTranslator` function, allowing developers to override some of the translations if they wish to do so.

macrozone commented:
> @macrozone Please see my latest commit.
> 
> 1. I have removed all default labels `|| 'text'`, they really weren't pretty.
> 2. `UiTranslator` type is moved to `options.ts` as requested.
> 3. changed `uiTranslator` prop from `(labels?: string) => string | undefined` to `(label:string) => string`. So that developers have to return a string if they are using TypeScript.
> 
> To recap:
> 
> 1. `uiTranslator` prop is optional, if not provided, `Editor` component default it to `label=>label` to return the label as-is. It has very little impact on the rest of the code.
> 2. English labels are used as keys. This assumes English labels are unique. If they are used in multiple places, their translations will be the same.
> 3. The signature `uiTranslator` prop will force developers to always return a string hopefully. If not, I guess they will find come blank labels quickly and fix the issue themselves.
> 
> We do have full control of the default `uiTranslator`. Later, if we have a complete set of translation for a language, it is very easy to introduce a `uiLanguage` prop to switch to other supported language. All we have to do is inserting dictionaries into the default `uiTranslator` function to return the correct translated labels instead of returning labels as-is. We could even accept a partial dictionary from the developer and hook it into the default `uiTranslator` function, allowing developers to override some of the translations if they wish to do so.

can you do like i proposed here https://github.com/react-page/react-page/pull/935#discussion_r606786813 ? I think that is the best compromise

macrozone commented on packages/editor/src/editor/Editor.tsx:
```diff
@@ -55,6 +56,7 @@ const Editor: React.FC<EditorProps> = ({
   lang = languages?.[0].lang ?? 'default',
   pluginsWillChange,
   cellSpacing,
+  uiTranslator = (label) => label,
```
as described here https://github.com/react-page/react-page/pull/935#discussion_r606786813

uiTranslator should not have a default `(label) => label`. Instead the {t} function should do that if the user provided uiTranslator does not return a value. This allows us to use default translations.

rexebin replied:
I considered your suggestion, a lot. The t function is being provided via a context and it is a singleton. The same t function is being used everywhere, so it cannot be specific to individual use cases. I don't know how we can implement a useDefaultTranslations function to fit all.

The closest I could think of is that the t function accepts a dictionary as a second argument, and use it in place of useDefaultTranslations. However It gives more problem than it solves and it doesn't work in some cases. For example, in the PluginButton component, it renders all plugin buttons, there is so dynamic, you cannot supply a default dictionary. 

English version labels are being used as key, returning the key as-is covers the lack of translations.

macrozone replied:
here is the t function i proposed:

```
// this is the t function returned by the useUiTranslator hook. 
const t = (key: string) => {
 const translated = userDefinedUiTranslatorFunction(key)
 if(translated === null){
   return defaultTranslatorFunction(key) // that can be the identity to begin with and later can use some default translation object
 }
}
```

`defaultTranslatorFunction would just be `(key) => key`

The problem is, that if we introduce some new label and the developers use an approach where they would not handle fallbacks, we end up with an empty label, which might affect usability negativly.

that's why i am all for handling this case on our side as it is super easy to do.

Your current solution does not do that. If the user provides a translator function that returns null for some keys, it will just not show the label at all.

rexebin replied:
If you spin up the dev server and go to i18n demo page, you will see what I mean. I only provided translations toolbar, and I return "label+(to translate)" for everything else to show all labels are being covered.

rexebin replied:
> here is the t function i proposed:
> 
> 
> 
> ```
> 
> // this is the t function returned by the useUiTranslator hook. 
> 
> const t = (key: string) => {
> 
>  const translated = userDefinedUiTranslatorFunction(key)
> 
>  if(translated === null){
> 
>    return defaultTranslatorFunction(key) // that can be the identity to begin with and later can use some default translation object
> 
>  }
> 
> }
> 
> ```
> 
> 
> 
> `defaultTranslatorFunction would just be `(key) => key`
> 
> 
> 
> The problem is, that if we introduce some new label and the developers use an approach where they would not handle fallbacks, we end up with an empty label, which might affect usability negativly.
> 
> 
> 
> that's why i am all for handling this case on our side as it is super easy to do.
> 
> 
> 
> Your current solution does not do that. If the user provides a translator function that returns null for some keys, it will just not show the label at all.
> 
> 
> 
> 

The signature of the uiTranslator prop require the developers to return a string. So they have to return a string, not a null or undefined. 

It is also very easy for them to always return key as-is as fallback. So they never see a blank label.

macrozone replied:
please implement the fallback in the hook function as described above. Typescript signatures are no runtime guarantee.
 Most developers will just do something like:

```
const uiTranslator = key => MY_TRANSLATION_DICT[key]

```

Is there a show stopper from doing it like that? i described it above in multiple places. Should i push a commit with it?

rexebin replied:
Please push a commit. I don't know how to implement your suggestions. I am keen to see how it works.

macrozone replied:
showed on a comment above

rexebin replied:
Oh, I think I understand now, are you saying if the given function returns null or undefined, the hook default to return key?

rexebin replied:
Do you want me to have one last go or you can take care of it from here?

Sorry for the mess..

macrozone replied:
yes! exactly, does this make sense? Sorry for being unclear and thank you for your effort!

macrozone replied:
sure, go on, i will merge asap, rest looks good!

macrozone commented on packages/editor/src/core/components/hooks/options.ts:
```diff
@@ -26,6 +26,15 @@ export const OptionsContext = createContext<Options>({
+/**
+ * @returns the an object with a single `t` function for ui translations
+ */
+export const useUiTranslator = () => useContext(UiTranslatorContext);
```
maybe this would work?

```tsx
useUiTranslator = () => {
  
  const userTranslatorFunciton =  useContext(UiTranslatorContext);
  return (key: string) => {
       return userTranslatorFunciton(key) ?? key
   }
}
````


```

macrozone replied:
@rexebin what do you think?

macrozone commented:
maybe i also did not show some comments of mine because of this weird review feature....

rexebin commented:
Done. I handled undefined and null case. Leaving empty string out, maybe in some rare cases some app wants a label to be empty.

macrozone commented:
:tada: This PR is included in version 2.0.0-beta.12 :tada:

The release is available on:
- `v2.0.0-beta.12`
- [GitHub release](https://github.com/react-page/react-page/releases/tag/v2.0.0-beta.12)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

macrozone commented:
:tada: This PR is included in version 2.0.0 :tada:

The release is available on:
- `v2.0.0`
- [GitHub release](https://github.com/react-page/react-page/releases/tag/v2.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
